src/hotspot/share/opto/optoreg.hpp

```diff
@@ -184,27 +185,56 @@ class OptoReg {
 class OptoRegPair {
   friend class VMStructs;
 private:
-  short _second;
-  short _first;
+  typedef short Name;
+  Name _second;
+  Name _first;
+
 public:
-  void set_bad (                   ) { _second = OptoReg::Bad; _first = OptoReg::Bad; }
-  void set1    ( OptoReg::Name n  ) { _second = OptoReg::Bad; _first = n; }
-  void set2    ( OptoReg::Name n  ) { _second = n + 1;       _first = n; }
-  void set_pair( OptoReg::Name second, OptoReg::Name first    ) { _second= second;    _first= first; }
-  void set_ptr ( OptoReg::Name ptr ) {
+  static constexpr bool can_fit(OptoReg::Name n) {
+    return n <= std::numeric_limits<Name>::max();
+  }
+  void set_bad() {
+    _second = OptoReg::Bad;
+    _first = OptoReg::Bad;
+  }
+  void set1(OptoReg::Name n) {
+    assert(can_fit(n), "overflow");
+    _second = OptoReg::Bad;
+    _first = n;
+  }
+  void set2(OptoReg::Name n) {
+    assert(can_fit(n + 1), "overflow");
+    assert(can_fit(n), "overflow");
+    _second = n + 1;
```

Comment:
I think we could just use checked_cast<>() here and then we don't need can_fit.  What do you think?

Reply:
I see we still need can_fit() for the static asserts, so nevermind.